File: Data Structures & Algorithms/decode-ways/submission-0.py
class Solution:
    
    def numDecodings(self, s: str) -> int:
        memo = {}
        def dp(ss:str) -> int:
            # if stored in memo
            if memo.get(ss):
                return memo[ss]
            # base case
            if ss == "":
                return 1
            if ss[0] == "0":
                return 0
            if len(ss) == 1:
                return 1
            else:
                # Recursive case for 1 and 2 (and not last digit)
                if ss[0] == "1" or ss[0] == '2':
                    if ss[0:2] == '10' or ss[0:2] == '20':
                        if len(ss) == 2:
                            memo[ss] = 1
                            return memo[ss]
                        else:
                            memo[ss] = dp(ss[2:])
                            return memo[ss]
                    elif int(ss[0:2]) <= 26:
                        memo[ss] = dp(ss[1:]) + dp(ss[2:])
                        return memo[ss]
                    else: # 27-29
                        memo[ss] = dp(ss[1:]) 
                        return memo[ss]

                # Recursive case for 3-9
                else:
                    memo[ss[1:]] = dp(ss[1:])
                    return memo[ss[1:]]
        return dp(s)

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.cpp
#include <algorithm>
class Solution {
public:
    int m;
    int maxProfit(vector<int>& prices) {
        m = prices.size();
        vector<vector<int>> memo(m+1, vector<int>(m, -1));
        return dp(prices, m, 0, memo);
    }

    int dp(vector<int>& prices, int buy_i, int pointer, vector<vector<int>>& memo) {
        // base case
        if (pointer >= m) {
            return 0;
        }
        
        if (memo[buy_i][pointer] != -1) {
            return memo[buy_i][pointer];
        }

        // recursive case: holding
        if (buy_i != m) {
            int sell_case = prices[pointer] - prices[buy_i] + dp(prices, m, pointer + 2, memo);
            int hold_case = dp(prices, buy_i, pointer + 1, memo);
            memo[buy_i][pointer] = (sell_case > hold_case) ? sell_case : hold_case;
            return memo[buy_i][pointer];
        } else {
            int buy_case = dp(prices, pointer, pointer + 1, memo);
            int wait_case = dp(prices, m, pointer + 1, memo);
            memo[buy_i][pointer] = (buy_case > wait_case) ? buy_case : wait_case;
            return memo[buy_i][pointer];
        }
    }
};
